app/modules/sale.js

```diff
@@ -1,58 +1,199 @@
 // @flow
 import { wallet, api } from 'neon-js'
+import { flatten } from 'lodash'
 
-import { showErrorNotification, showInfoNotification, showSuccessNotification } from './notifications'
-import { getWIF, LOGOUT } from './account'
+import {
+  showErrorNotification,
+  showInfoNotification,
+  hideNotification
+} from './notifications'
+import {
+  getWIF,
+  LOGOUT,
+  getAddress,
+  getIsHardwareLogin,
+  getSigningFunction,
+  getPublicKey
+} from './account'
 import { getNetwork } from './metadata'
-import { getNEO } from './wallet'
-import { toBigNumber, toNumber } from '../core/math'
+import { getNEO, getGAS } from './wallet'
+
+import { toNumber } from '../core/math'
 import asyncWrap from '../core/asyncHelper'
+import { ASSETS } from '../core/constants'
+import {
+  validateMintTokensInputs,
+  validateOldMintTokensInputs
+} from '../core/sale'
+import { oldMintTokens } from '../core/oldMintTokens'
 
-// TODO: Rewrite this function
+const MESSAGES = {
+  PARTICIPATION_FAILED: 'Sale participation failed, please check your script hash again.'
+}
 
-export const participateInSale = (neoToSend: string, scriptHash: string) => async (dispatch: DispatchType, getState: GetStateType) => {
+export const participateInSale = (
+  neoToSend: string,
+  gasToSend: string,
+  scriptHash: string,
+  gasCost: string = '0'
+) => async (dispatch: DispatchType, getState: GetStateType) => {
   const state = getState()
   const wif = getWIF(state)
-  const NEO = getNEO(state)
+  const publicKey = getPublicKey(state)
+  const NEO = toNumber(getNEO(state))
+  const GAS = toNumber(getGAS(state))
```

Comment:
We should use `toBigNumber` on these, or just pass the string through to any internal functions we call.